tests/system.rs

```diff
@@ -0,0 +1,214 @@
+// output must match exactly.
+// FIXME(#28) would be good to check for error messages and fail on them, or at least report.
+#[test]
+fn system_tests() {
+    // Get all files in the tests/target directory
+    let files = fs::read_dir("tests/target").ok().expect("Couldn't read dir 1.");
+    let files = files.chain(fs::read_dir("tests").ok().expect("Couldn't read dir 2."));
+    let files = files.chain(fs::read_dir("src/bin").ok().expect("Couldn't read dir 3."));
+    // turn a DirEntry into a String that represents the relative path to the file
+    let files = files.map(get_path_string);
+    // hack because there's no `IntoIterator` impl for `[T; N]`
+    let files = files.chain(Some("src/lib.rs".to_owned()).into_iter());
+
+    let (count, fails) = check_files(files);
+
+    // Display results
+    println!("Ran {} idempotent tests.", count);
+    assert!(fails == 0, "{} idempotent tests failed", fails);
+
+    // Get all files in the tests/source directory
+    let files = fs::read_dir("tests/source").ok().expect("Couldn't read dir 4.");
+    // turn a DirEntry into a String that represents the relative path to the file
+    let files = files.map(get_path_string);
+
+    let (count, fails) = check_files(files);
+
+    // Display results
+    println!("Ran {} system tests.", count);
+    assert!(fails == 0, "{} system tests failed", fails);
+}
+
+// For each file, run rustfmt and collect the output.
+// Returns the number of files checked and the number of failures.
+fn check_files<I>(files: I) -> (u32, u32)
+    where I: Iterator<Item = String>
+{
+    let mut count = 0;
+    let mut fails = 0;
+
+    for file_name in files.filter(|f| f.ends_with(".rs")) {
+        println!("Testing '{}'...", file_name);
+        match idempotent_check(file_name) {
+            Ok(()) => {},
+            Err(m) => {
+                print_mismatches(m);
+                fails += 1;
+            },
+        }
+        count += 1;
+    }
+
+    (count, fails)
+}
+
+fn print_mismatches(result: HashMap<String, String>) {
+    for (_, fmt_text) in result {
+        println!("{}", fmt_text);
+    }
+}
+
+// Ick, just needed to get a &'static to handle_result.
+static HANDLE_RESULT: &'static Fn(HashMap<String, String>) = &handle_result;
+
+pub fn idempotent_check(filename: String) -> Result<(), HashMap<String, String>> {
+    let config = get_config(&filename);
+    let args = vec!["rustfmt".to_owned(), filename];
+    // this thread is not used for concurrency, but rather to workaround the issue that the passed
+    // function handle needs to have static lifetime. Instead of using a global RefCell, we use
+    // panic to return a result in case of failure. This has the advantage of smoothing the road to
+    // multithreaded rustfmt
+    thread::catch_panic(move || {
+        run(args, WriteMode::Return(HANDLE_RESULT), &config);
+    }).map_err(|any|
+        *any.downcast().ok().expect("Downcast failed.")
```

Comment:
wherever that `unwrap` is happening in rustfmt, i think we need an `expect` there xD